### PR TITLE
start from interruption

### LINKED ISSUE #34: Add start-from-interruption logic
### Goals and reasoning

During the call we identified requirements:

- This feature is a must for all micro-indexers
- It should work for any run scenario:
  - Follow the tip of the network
  - First-time indexing
  - Filling the indexing gaps
  - Reindexing stuff
- Has to be reliable

**NB!** Such a feature is required by almost every indexer out there in a wild, but can’t be the same for each of them because it depends on the work an indexer does. That’s why it is not provided as a part of Lake Framework and has to be reinvented most of the time.

We have observed a few approaches this feature might be achieved with.

---

### TL;DR: TODO: Implement the Redis approach

---

### Naive approach

We can store the last processed block height on the filesystem near the binary and read that value to start from that block in case an indexer was stopped. The implementation of this approach is described in the [tutorial for Lake Indexers](https://docs.near.org/tutorials/indexer/lake-start-options#frominterruption)

Pros:
- Doesn't introduce new dependencies
- Relatively simple to implement
- Straightforward 
- Can be hacked on the fly (have no idea what for, but it can)
Cons:
- Running in Docker means the filesystem will be reset
- In some cases, the indexer might be interrupted during the file update and results might be unpredictable

### SQL approach

Since we save data to the database. We can query the database to get the latest recorded block height (or a min from two max block heights when we write to a couple of tables) and start from that.

Pros:
- No new dependency
- Relatively reliable for:
  - “follow the tip of the network” run scenario 
  - “first indexer” run scenario
Cons:
- Doesn’t work for:
  - “filling the gaps” scenario out of the box (requires giving the indexing range to determine where it was interrupted)
  - “reindex” scenario (in this case it would start from the latest possible block in case of interruption, might be tweaked though)

### Redis approach

The idea is to have an instance of Redis for the EAPI project, all the micro-indexers have an access to it and write the last processed block height on their own to their Redis key.

Pros:
- Proved to be a reliable solution (see [Alertexer](https://github.com/near/alertexer/blob/main/storage/src/lib.rs#L119-L126) for example)
- Quick and relatively easy to implement

Cons:
- Adds [Redis](https://crates.io/crates/redis) dependency, though it’s 139 kB
- Requires a managed Redis instance on GCP and DevOps magic to set up access for each current and future micro-indexer in the project (@ecp88)

### Decision

The decision was made during the call to go with **the Redis approach**. I’ve decided to mention other discussed approaches for the history (and to avoid questions like “why didn’t you just do…”, because we thought and decided not to 😅)

### PR DESCRIPTION
**Pull request by roshaans**
Start from interruption similarly to how it is implemented in `near-indexer-balances`

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -28,12 +28,21 @@ pub struct AccountWithContract {
 async fn main() -> anyhow::Result<()> {
     dotenv().ok();
     let opts: Opts = Opts::parse();
+    let _worker_guard = init_tracing(opts.debug)?;
```
Comment (roshaans) [inline src/main.rs:31]: Moved this up here to be consistent with the order in `near-indexer-balances`

File: src/models/mod.rs [hunk 2]
```diff
@@ -115,3 +117,71 @@ pub(crate) fn create_placeholder(
...
+                }
+            }
+        }
+    }
+}
+
+pub(crate) async fn start_after_interruption(
+    pool: &sqlx::Pool<sqlx::Postgres>,
+) -> anyhow::Result<u64> {
+    let query = "SELECT MAX(block_height)
```
Comment (roshaans) [inline src/models/mod.rs:167]: Union the NFT_EVENTS and COIN_EVENTS tables together and finds the max block height from those two.

Comment (telezhnaya) [reply src/models/mod.rs:167]: Why max block height?

File: src/models/mod.rs [hunk 3]
```diff
@@ -115,3 +117,71 @@ pub(crate) fn create_placeholder(
...
+                        FROM (
+                            SELECT block_height, block_timestamp
+                            FROM coin_events
+                            UNION ALL
+                            SELECT block_height, block_timestamp
+                            FROM nft_events
+                        ) AS combined_events
+                        GROUP BY block_timestamp
+                        ORDER BY block_timestamp desc
+                        LIMIT 1";
```
Comment (telezhnaya) [inline src/models/mod.rs:177]: Have you tested the speed of this query?

**Review by telezhnaya — CHANGES_REQUESTED** [review]

File: src/models/mod.rs (hunk 2 above)
Comment (roshaans) [reply src/models/mod.rs]: We should actually take MIN here. 

If in a given block height, the indexer has to write to both nft_events and coin_events and the indexer halts between the independent writes to the database. We should choose the minimum of the two block heights to catch this edge case although we end up indexing from the previous 1000 blocks, it is better to be correct. Thanks :) Just updated to take the MIN.

File: src/models/mod.rs [hunk 4]
```diff
@@ -115,3 +117,67 @@ pub(crate) fn create_placeholder(
...
+pub(crate) async fn start_after_interruption(
+    pool: &sqlx::Pool<sqlx::Postgres>,
+) -> anyhow::Result<u64> {
+    // Gets the highest block height from both the nft_events and coin_events table
+    let query = "SELECT MIN(block_height)
+                       FROM (SELECT block_height
+                            FROM   coin_events
+                            UNION ALL
+                            SELECT block_height
+                            FROM   nft_events) AS combined_heights";
```
Comment (telezhnaya) [inline src/models/mod.rs:173]: Try to create samples and test it

**Review by telezhnaya — CHANGES_REQUESTED** [review]

File: src/models/mod.rs (hunk 4 above)
Comment (roshaans) [reply src/models/mod.rs:173]: I test this in the prod database  

The query runs on average 134 ms with ~692,980,032 million rows inside of coin_events at the moment. 

https://github.com/near/near-indexer-events/pull/37#discussion_r1048697119

Comment (telezhnaya) [reply src/models/mod.rs:173]: I meant functional testing.

I can't ask you to write unit tests here, it's cruel, we don't have any testing environment set up.
Let's imagine we have one.
Which tests do you prefer to cover this function? Could you please draft them?
You don't need to write the code, just create a text draft with the examples, what's the input (from the DB) and what output should we have.

**Pull request closed without merging**